### PR TITLE
CLI training and config

### LINKED ISSUE #257: training with CLI
It's on the TODO to build out the CLI to incorporate training with a config file.

- How should the config file look?
- Should a config be created via a separate command?
- How to handle how a dataset should be used?
- Should it support accelerate?

### PR DESCRIPTION
**Pull request by TheFusion21**
added command

`imagen train` trains a unet from a given config file
`imagen config` generates basic config file

- support train and validation
- most if not all parameter for Imagen/ElucidatedImagen Config
- sample between epochs

closes #257

### CODE REVIEW

**Comment by lucidrains** [comment]
@TheFusion21 Hey Kay! Thank you for this PR! Do you want to write up a small section in the Readme for how to use?

Does this support resuming training as well?

File: imagen_pytorch/cli.py [hunk 1]
```diff
@@ -50,3 +59,158 @@ def imagen(
 
     print(f'image saved to {str(image_path)}')
     return
+
+@imagen.command(help = 'Generate a config for the Imagen model')
+def config():
+    config = {
```
Comment (lucidrains) [inline imagen_pytorch/cli.py:65]: could this be placed in a `default_config.json` and read from that file?

File: imagen_pytorch/cli.py [hunk 2]
```diff
@@ -4,23 +4,32 @@
...
 
 def main():
     pass
 
-@click.command()
+@click.group()
+def imagen():
+    pass
+
+@imagen.command(help = 'Sample from the Imagem model checkpoint')
```
Comment (lucidrains) [inline imagen_pytorch/cli.py:27]: small typo `imagem`

File: imagen_pytorch/cli.py [hunk 3]
```diff
@@ -50,3 +59,158 @@ def imagen(
...
+        config_data = json.loads(f.read())
+
+    assert 'checkpoint_path' in config_data, 'checkpoint path not found in config'
+    
+    model_path = Path(config_data['checkpoint_path'])
+    full_model_path = str(model_path.resolve())
+    
+    # setup imagen config
+
+    if config_data['type'] == 'elucidated':
```
Comment (lucidrains) [inline imagen_pytorch/cli.py:149]: this could be shortened to

```python
imagen_config_klass = ElucidatedImagenConfig if config_data['type'] == 'elucidated' else ImagenConfig
imagen = imagen_config_klass(**config_data['imagen']).create()
```

File: imagen_pytorch/data.py [hunk 4]
```diff
@@ -23,7 +31,73 @@ def convert_image_to(img_type, image):
...
+            images.append(image)
+
+        if len(masks) == 0:
+            return None
+
+        
+        max_len = max([masks[i].shape[1] for i in range(len(masks))])
+
+        newbatch = []
+        for i in range(len(masks)):
```
Comment (lucidrains) [inline imagen_pytorch/data.py:75]: https://pytorch.org/docs/stable/generated/torch.nn.utils.rnn.pad_sequence.html

are you doing something similar to this function in these 10-15 loc?

File: imagen_pytorch/data.py [hunk 5]
```diff
@@ -23,7 +31,73 @@ def convert_image_to(img_type, image):
...
+                    image = self.fetch_single_image(item[self.url_label])
+                else:
+                    image = item[self.image_label]
+                image = self.transform(image.convert('RGB'))
+            except:
+                continue
+
+            text, mask = t5.t5_encode_text([item[self.text_label]], return_attn_mask=True, name=self.name)
+            texts.append(text)
+            masks.append(mask)
```
Comment (lucidrains) [inline imagen_pytorch/data.py:65]: you don't actually have to manage the `mask` anymore, as the framework automatically derives it from the text embeddings https://github.com/lucidrains/imagen-pytorch/blob/main/imagen_pytorch/imagen_pytorch.py#L2201

File: imagen_pytorch/data.py (hunk 4 above)
Comment (TheFusion21) [reply imagen_pytorch/data.py:75]: the default collator crashes if the tensors are not the same length.
don't know how to handle it differently

File: imagen_pytorch/data.py (hunk 5 above)
Comment (TheFusion21) [reply imagen_pytorch/data.py:65]: From my trainings the result are different when I don't use masks

**Comment by TheFusion21** [comment]
> @TheFusion21 Hey Kay! Thank you for this PR! Do you want to write up a small section in the Readme for how to use?
> 
> Does this support resuming training as well?

will update readme.
Yes it does allow for resuming training
Only problem right now is that I can't use workers because imagen and t5 are both on the same device and not on the same thread

**Comment by TheFusion21** [comment]
FYI this should also work with accelerate but I can't verify
```bash
$ accelerate launch --no_python imagen train
```

**Comment by lucidrains** [comment]
@TheFusion21 graciously accept! thanks for getting the ball rolling!
